numpy/ma/core.py

```diff
@@ -6309,6 +6309,18 @@ def copy(self, *args, **kwargs):
         # precedent for this with `np.bool_` scalars.
         return self
 
+    def __setattr__(self, attr, value):
+        if not self.__has_singleton():
```

Comment:
Can we combine the two to `if self.__has_singleton() and self is self.__singleton`? would imply longer comment but no duplicated code.

Reply:
I wrote it this way because the justification for the two cases is different, and so the code duplication is perhaps justifiable - it may be desirable to only allow the `__class__` attribute to be set in the last branch, for instance.

Reply:
OK, no big deal.